### ISSUE TITLE
[SETUP ISSUE] wp-browser v4 wp-config file parsing issue

### ISSUE THREAD

**Issue by snandos** [lucatume/wp-browser#627]
**Environment**
OS: Docker php:8.1-apache
PHP version: 8.1
Installed Codeception version: 5.0.11
Installed wp-browser version: v4.x-dev 3b12879
WordPress version: 6.3
Local development environment: Docker
WordPress structure and management: Custom Docker-Image for Development

**Did you use the `codecept init wpbrowser` command?**
yes

**Did you take a look at Codeception and wp-browser documentation?**
yes

**Codeception configuration file**
```yaml
actor: Tester
paths:
  tests: tests
  log: tests/output
  data: tests/data
  support: tests/support
  envs: tests/envs
  output: tests/output
  
params:
  - tests/.test.env    
  - CODECEPTION_ENV: 'true'
```

**Suite configuration file**

```yaml
actor: UnitTester
bootstrap: _bootstrap.php
modules:
    enabled:
        - lucatume\WPBrowser\Module\WPLoader
    config:
        lucatume\WPBrowser\Module\WPLoader:
           loadOnly: true
           wpRootFolder: "%WORDPRESS_ROOT_DIR%" 
           dbUrl: "%WORDPRESS_DB_URL%"
           wpDebug: true
           tablePrefix: "%TEST_TABLE_PREFIX%"
           domain: "%WORDPRESS_DOMAIN%"
           adminEmail: "admin@%WORDPRESS_DOMAIN%"
           title: "Integration Tests"
           plugins: [""]
           theme: "my-theme" 
```

**Describe the issue you're encountering**
I want to unit test my WordPress setup with codeception. I'm already doing acceptance tests successfully (without wp-browser), so I would love to use codeception for unittests as well.
All my code resides in my custom theme. For development I use a custom written Dockerfile derived from the official WordPress-Dockerfile. Codeception is being run inside this docker container.
Because of the docker setup, I need to use a wp-config file which is not default. For example I use a `getenv_docker` function to retrieve environment variables either from a .env file or from environment variables itself. Like `define( 'DB_NAME', getenv_docker('WORDPRESS_DB_NAME', 'wordpress') );`.

Because of this, wp-browser won't do anything since it is not able to parse my wp-config:
```
In WPConfigFile.php line 132:                                           
  Could not parse the wp-config.php file:                                             

In WPConfigFile.php line 118:                                                       
  [lucatume\WPBrowser\Process\ProcessException (255)]  
                                                       
In SerializableClosure.php line 155:                                                                 
  Uncaught Exception: Serialization of 'Closure' is not allowed 
```

When uncommenting all these "dynamic" defines, at least it can parse the file. But of course won't be able to read the needed configs:
```
In WPConfigFile.php line 164:                                 
  Constant DB_NAME not defined.  
```

My question is: why does it need to parse and read an existing `wp-config.php` file in the first place, when I define those options in the codeception scenario? Am I missing something in the config? The `loadOnly` option is still confusing for me.

If the parsing of a wp-config is inevitable, do you have an idea how I can get around this issue while still keeping my wp-config in place? Like specifying a different wp-config file.

Thank you for your awesome work!

**Comment by lucatume** [comment]
Hi @snandos,

version 4 is still not out, but I appreciate you taking it for a ride.

> why does it need to parse and read an existing wp-config.php file in the first place, when I define those options in the codeception scenario? Am I missing something in the config?

It does not need to: it does read your `wp-config.php` file because you're using the module with `loadOnly: true`.
[Read more about it here](https://github.com/lucatume/wp-browser/blob/v4/docs/modules/WPLoader.md#wploader-module). 
If you set `loadOnly: false`, and correctly configure the module, it should solve your problems.

**Comment by snandos** [comment]
Hi @lucatume,

thanks for your response!
I use the dev version 4 because of some other dependencies for codeception 5. Thank you for upgrading wp-browser to be compatible! 😊

After more testing I learned, that as soon as a `wp-config.php` file is found within the WordPress root directory or one directory above, the file will get parsed, no matter if `loadOnly` is set to `false` or not because of the logic within `findWpConfigFilePath`.

So I had to fix my wp-config file. Got it working now to the point where an empty test is run 👍 

Unfortunately I get errors when including woocommerce, because it looks like the woocommerce classes are being serialized and unserialized which is not supported.

```
[PHPUnit\Framework\Error\Notice] Function __wakeup was called <strong>incorrectly</strong>. Unserializing instances of this class is forbidden.
Backtrace:
include('/themes/theme-name/vendor/codeception/codeception/codecept'),
require('/themes/theme-name/vendor/codeception/codeception/app.php'), {closure},
Codeception\Application->run,
Symfony\Component\Console\Application->run,
Symfony\Component\Console\Application->doRun,
Symfony\Component\Console\Application->doRunCommand,
Symfony\Component\Console\Command\Command->run,
lucatume\WPBrowser\Command\RunAll->execute,
Codeception\Command\Run->execute,
Codeception\Command\Run->runSuites,
Codeception\Codecept->run,
Codeception\Codecept->runSuite,
Codeception\SuiteManager->run,
Codeception\Suite->run,
Codeception\Test\Test->realRun,
Codeception\Test\TestCaseWrapper->test,
PHPUnit\Framework\TestCase->runBare,
PHPUnit\Framework\TestCase->snapshotGlobalState,
PHPUnit\Framework\TestCase->createGlobalStateSnapshot,
SebastianBergmann\GlobalState\Snapshot->__construct,
SebastianBergmann\GlobalState\Snapshot->snapshotGlobals,
unserialize,
WooCommerce->__wakeup,
wc_doing_it_wrong Please see <a href="https://wordpress.org/documentation/article/debugging-in-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version 2.1.) 
```

Is this still in the scope of your codeception addon? My guess is, that this has something to do with some sort of process isolation of PHPUnit, but I'm quite a noob in this area :)

**Comment by snandos** [comment]
Hi,

I've created the following demo repo: https://github.com/snandos/wp-browser-wakeup-error-repo
Hope this helps!

**Comment by lucatume** [comment]
@snandos, thanks for setting up the repository and providing a good way to reproduce the issue. Much appreciated.

I've fixed the issue in version `4.0.2` and tested it out on your branch.

A note: you might have done that while trying to debug the issue, [but the `setUp` method should call the `parent::setUp` method to work correctly](https://github.com/snandos/wp-browser-wakeup-error-repo/blob/master/tests/unit/SimpleUnitTest.php#L11-L12); this was **not** the cause of the issue.

The cause of the issue was the fact that, differently from version 3 of the project, in version 4 I'm leveraging PHPUnit global state, and static attribute, handling between tests. [The `WPTestCase` is setting up  a list of global variables and static attributes that should not be backed up and restored between tests](https://github.com/lucatume/wp-browser/blob/1f6a3a9b19150bc9c0d62d74a9a13f507285df4b/src/TestCase/WPTestCase.php#L13-L52); WooCommerce specific classes and attributes were excluded and their attempted backup was causing the issue.
